src/x509.cpp

```diff
@@ -31,17 +34,26 @@ using namespace openssl;
 
 X509Certificate X509Certificate::fromPEM(const std::string &pem)
 {
+    std::string formattedPem = util::reformatPEMCertificate(pem);
     BioObject bio{BioObject::Types::MEM};
-    bio.write(pem);
+    bio.write(formattedPem);
     auto cert = _PEM_read_bio_X509(bio.internal());
     return X509Certificate{std::move(cert)};
 }
 
 X509Certificate X509Certificate::fromPEMFile(const std::string &filename)
 {
-    FileBio bio{filename, FileBio::FileMode::READ, FileBio::FileType::TEXT};
-    auto cert = _PEM_read_bio_X509(bio.internal());
-    return X509Certificate{std::move(cert)};
+    // Change to be reverted to previous solution after OpenSSL parsing
+    // bug for PEM certificates has been fixed.
+    std::ifstream pemFile(filename);
+    std::stringstream buffer;
+    if (pemFile) {
+        buffer << pemFile.rdbuf();
```

Comment:
I'm not sure if we should add error handling here.
pemFile should be at eof after this but the bad bit should not be set?

Reply:
right.. but for some reason error handling is a bit weird http://www.cplusplus.com/reference/ostream/ostream/operator%3C%3C/:

(2) stream buffers
    Retrieves as many characters as possible from the input sequence controlled by the stream buffer object pointed by sb (if any) and inserts them into the stream, until either the input sequence is exhausted or the function fails to insert into the stream.
    Internally, the function accesses the output sequence by first constructing a sentry object. Then (if good), it inserts characters into its associated stream buffer object as if calling its member function sputc, and finally destroys the sentry object before returning.

failbit | The function failed to format val (it may also be set if the construction of sentry failed). For (2), it is set if no characters could be extracted from the object pointed by sb.
-- | --
badbit | Either  the insertion on the stream failed, or some other error happened (such  as when this function catches an exception thrown by an internal  operation). When set, the integrity of the stream may have been affected. For (2), it is also set if sb is a null pointer.

So i guess checking for fail() on the buffer should be sufficient, or am I missing something

Reply:
unclear. It reads to me like both indicate different error scenarios but in principle I agree that error handling with the stream APIs is weird. Maybe we should better be safe than sorry and check both?

Reply:
`fail()` actually checks both bits:
>  Returns true if either (or both) the failbit or the badbit error state flags is set for the stream.
http://www.cplusplus.com/reference/ios/ios/fail/

Reply:
or did you mean on both buffers?